src/machine/machined-varlink.c

```diff
@@ -392,56 +392,52 @@ static int vl_method_get_memberships(sd_varlink *link, sd_json_variant *paramete
-                return 0;
-
-        assert(addresses);
+        assert(addresses || n_addresses == 0);
         assert(ret);
 
         FOREACH_ARRAY(a, addresses, n_addresses) {
-                _cleanup_(sd_json_variant_unrefp) sd_json_variant *entry = NULL;
-                r = sd_json_buildo(
-                                &entry,
+                r = sd_json_variant_append_arraybo(
+                                &array,
                                 JSON_BUILD_PAIR_UNSIGNED_NON_ZERO("ifindex", a->ifindex),
                                 SD_JSON_BUILD_PAIR_INTEGER("family", a->family),
                                 SD_JSON_BUILD_PAIR_BYTE_ARRAY("address", &a->address.bytes, FAMILY_ADDRESS_SIZE(a->family)));
                 if (r < 0)
                         return r;
-
-                r = sd_json_variant_append_array(ret, entry);
-                if (r < 0)
-                        return r;
         }
 
+        *ret = TAKE_PTR(array);
         return 0;
 }
 
 static int list_machine_one_and_maybe_read_metadata(sd_varlink *link, Machine *m, bool more, AcquireMetadata am) {
-        _cleanup_(sd_json_variant_unrefp) sd_json_variant *addr_array = NULL;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL, *addr_array = NULL;
         _cleanup_strv_free_ char **os_release = NULL;
         uid_t shift = UID_INVALID;
-        int r, n = 0;
+        int r;
 
         assert(link);
         assert(m);
 
-        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        if (!sd_varlink_is_processing_method(link))
+                return 0;
 
         if (should_acquire_metadata(am)) {
                 _cleanup_free_ struct local_address *addresses = NULL;
-                n = machine_get_addresses(m, &addresses);
-                if (n < 0 && am == ACQUIRE_METADATA_GRACEFUL)
-                        log_debug_errno(n, "Failed to get address (graceful mode), ignoring: %m");
-                else if (n == -ENONET)
+
+                r = machine_get_addresses(m, &addresses);
+                if (r < 0 && am == ACQUIRE_METADATA_GRACEFUL)
+                        log_debug_errno(r, "Failed to get address (graceful mode), ignoring: %m");
+                else if (r == -ENONET)
                         return sd_varlink_error(link, "io.systemd.Machine.NoPrivateNetworking", NULL);
-                else if (ERRNO_IS_NEG_NOT_SUPPORTED(n))
+                else if (ERRNO_IS_NEG_NOT_SUPPORTED(r))
                         return sd_varlink_error(link, "io.systemd.Machine.NotAvailable", NULL);
-                else if (n < 0)
-                        return log_debug_errno(n, "Failed to get addresses: %m");
+                else if (r < 0)
+                        return log_debug_errno(r, "Failed to get addresses: %m");
                 else {
-                        r = json_build_local_addresses(addresses, n, &addr_array);
+                        r = json_build_local_addresses(addresses, r, &addr_array);
```

Comment:
I found this change confusing. You use 'r,' which is an error code, as a "number of elements." I think using `n` is more explicit and easier to understand.

Reply:
But, both integers... I'd like to check error code in the above with `r` rather than `n`.

BTW, in the long run, I'd like to make functions that returns size of an array take another `size_t *ret_size` output parameter, instead of returning the size.